core/src/main/java/com/linecorp/armeria/internal/common/stream/PrependingPublisher.java

```diff
@@ -0,0 +1,144 @@
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.stream.NoopSubscriber;
+
+public final class PrependingPublisher<T> implements Publisher<T> {
+
+    private final T first;
+    private final Publisher<? extends T> rest;
+
+    public PrependingPublisher(T first, Publisher<? extends T> rest) {
+        this.first = first;
+        this.rest = rest;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        final RestSubscriber<T> restSubscriber = new RestSubscriber<>(first, subscriber);
+        rest.subscribe(restSubscriber);
+    }
+
+    static final class RestSubscriber<T> implements Subscriber<T>, Subscription {
+
+        private static final AtomicLongFieldUpdater<RestSubscriber> demandUpdater =
+                AtomicLongFieldUpdater.newUpdater(RestSubscriber.class, "demand");
+        private final T first;
+        private volatile boolean firstSent;
+        private Subscriber<? super T> downstream;
+        private volatile long demand;
+        @Nullable
+        private volatile Subscription upstream;
+        @Nullable
+        private volatile Throwable upstreamCause;
+        private volatile boolean upstreamCompleted;
+        private volatile boolean completed;
+
+        RestSubscriber(T first, Subscriber<? super T> downstream) {
+            this.first = first;
+            this.downstream = downstream;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T t) {
+            requireNonNull(t, "element");
+            demandUpdater.decrementAndGet(this);
+            downstream.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            requireNonNull(t, "throwable");
+            // delay onError until the first piece is sent
+            if (!firstSent) {
+                upstreamCause = t;
```

Comment:
I think this is not thread-safe.
Let's say thread A executes line 85 and just before executing line 86.
And thread B executes 115~130. Then we are going to miss this onError event.

To solve this simply, I think we can just call `downstream.onNext(first)` in `onNext()` or `onComplete()` of `RestSubscriber`. (in `onError()` I think we don't pass the `first`)
The `request(...)` method will just handle adding up the demand.

Reply:
I found out that it's a bit trickier to implement this because:
- There's a chance that `Subscriber.onComplete()` and `Subscription.request()` is called at the same time.
  - In that case, `onNext(first) should always be called before onComplete()`.

To solve this, I think we need to
- Use a queue that stores the elements in order.
- Or, subscribe to the `rest` after the `first` element is passed.
Either way, it's fine for me. So it's your choice. 😄
I tried the second approach though.
https://gist.github.com/minwoox/8ceb11c72846b9f5eff791b399972a91
(I didn't clean up the code. 😅 )

Reply:
> Or, subscribe to the rest after the first element is passed.

👍